fpmsyncd/bgpmon.py

```diff
@@ -0,0 +1,176 @@
+            output = commands.getoutput(cmd)
+            peer_info = json.loads(output)
+            # no exception, safe to Clean the "new" lists/dic for new sanpshot
+            del self.new_ipv4_n[:]
+            self.new_ipv4_n_state.clear()
+            del self.new_ipv6_n[:]
+            self.new_ipv6_n_state.clear()
+            if "ipv4Unicast" in peer_info and "peers" in peer_info["ipv4Unicast"]:
+                self.new_ipv4_n = peer_info["ipv4Unicast"]["peers"].keys()
+                for i in range (0, len(self.new_ipv4_n)):
+                    self.new_ipv4_n_state[self.new_ipv4_n[i]] = \
+                    peer_info["ipv4Unicast"]["peers"][self.new_ipv4_n[i]]["state"]
+
+            if "ipv6Unicast" in peer_info and "peers" in peer_info["ipv6Unicast"]:
+                self.new_ipv6_n = peer_info["ipv6Unicast"]["peers"].keys()
+                for i in range (0, len(self.new_ipv6_n)):
+                    self.new_ipv6_n_state[self.new_ipv6_n[i]] = \
+                    peer_info["ipv6Unicast"]["peers"][self.new_ipv6_n[i]]["state"]
+
+        except Exception:
+            syslog.syslog(syslog.LOG_ERR, "*ERROR* get_all_neigh_states Exception: %s"
+                    % (traceback.format_exc()))
+
+    def update_neigh_states(self):
+        # handle IPV4 case
+        for i in range (0, len(self.new_ipv4_n)):
+            neighb = self.new_ipv4_n[i]
+            key = "NEIGH_STATE_TABLE|%s" % neighb
+            if neighb in self.ipv4_n:
+                # only update the entry if state changed
+                if self.ipv4_n_state[neighb] != self.new_ipv4_n_state[neighb]:
+                    # state changed. Update state DB for this entry
+                    state = self.new_ipv4_n_state[neighb]
+                    self.db.set(self.db.STATE_DB, key, 'state', state)
+                    self.ipv4_n_state[neighb] = state
+                # remove this neighbor from old list since it is accounted for
+                self.ipv4_n.remove(neighb)
+            else:
+                # New neighbor found case. Add to dictionary and state DB
+                state = self.new_ipv4_n_state[neighb]
+                self.db.set(self.db.STATE_DB, key, 'state', state)
+                self.ipv4_n_state[neighb] = state
+        # Check for stale state entries to be cleaned up
+        while len(self.ipv4_n) > 0:
+            # remove this from the stateDB and the current nighbor state entry
+            neighb = self.ipv4_n.pop(0)
+            del_key = "NEIGH_STATE_TABLE|%s" % neighb
+            self.db.delete(self.db.STATE_DB, del_key)
+            del self.ipv4_n_state[neighb]
+        # Save the new List
+        self.ipv4_n = self.new_ipv4_n[:]
+
+        # handle IPV6 case
+        for i in range (0, len(self.new_ipv6_n)):
+            neighb = self.new_ipv6_n[i]
+            key = "NEIGH_STATE_TABLE|%s" % neighb
+            if neighb in self.ipv6_n:
+                # only update the entry if sate changed
```

Comment:
>sate [](start = 43, length = 4)

Typo again. This happens to be a good argument for code reuse, so if you fix in one part, you fix every part.

Reply:
@qiluo-msft Yes. I am removing the unnecessary duplication with my next check in.

Reply:
@qiluo-msft Code reuse completed.